File: server/usecases/AuthUseCase.ts
import { IAdminRepository } from "../repositories/interfaces";
import { verifyPassword, hashPassword } from "../utils/password";
import { signToken } from "../utils/jwt";
import { Admin } from "../models/Admin";
import { UnauthorizedError, NotFoundError, ValidationError } from "../utils/errors";
import { logger } from "../utils/logger";

export class AuthUseCase {
  constructor(private adminRepo: IAdminRepository, private jwtSecret: string) {}

  async login(username: string, passwordPlain: string): Promise<{ token: string; admin: Omit<Admin, "password"> }> {
    const admin = await this.adminRepo.findByUsername(username);
    if (!admin || !admin.password) {
      logger.warn(`Login failed - user not found: ${username}`, "AuthUseCase");
      throw new UnauthorizedError("Invalid username or password");
    }

    const isValid = await verifyPassword(passwordPlain, admin.password);
    if (!isValid) {
      logger.warn(`Login failed - invalid password for user: ${username}`, "AuthUseCase");
      throw new UnauthorizedError("Invalid username or password");
    }

    // Generate token
    const token = await signToken({ id: admin.id, username: admin.username }, this.jwtSecret);
    
    // Exclude password from return
    const { password, ...adminInfo } = admin;

    return {
      token,
      admin: adminInfo,
    };
  }

  async getProfile(adminId: number): Promise<Omit<Admin, "password">> {
    logger.debug(`Fetching profile for admin id=${adminId}`, "AuthUseCase");

    const admin = await this.adminRepo.findById(adminId);
    if (!admin) {
      logger.warn(`Profile not found for admin id=${adminId}`, "AuthUseCase");
      throw new NotFoundError("Admin not found");
    }

    const { password, ...adminInfo } = admin;
    return adminInfo;
  }

  async changePassword(adminId: number, oldPasswordPlain: string, newPasswordPlain: string): Promise<void> {
    const admin = await this.adminRepo.findById(adminId);
    if (!admin || !admin.password) {
      logger.warn(`Password change failed - admin not found: id=${adminId}`, "AuthUseCase");
      throw new NotFoundError("Admin not found");
    }

    const isValid = await verifyPassword(oldPasswordPlain, admin.password);
    if (!isValid) {
      logger.warn(`Password change failed - incorrect current password for admin id=${adminId}`, "AuthUseCase");
      throw new ValidationError("Incorrect current password");
    }

    const newHashed = await hashPassword(newPasswordPlain);
    await this.adminRepo.updatePassword(adminId, newHashed);
  }
}

File: checker2-go/internal/module/auth/usecase/auth_usecase.go
package usecase

import (
	"context"
	"fmt"

	"github.com/hidessh99/sub-cf-vpn/checker2-go/internal/infrastructure/config"
	"github.com/hidessh99/sub-cf-vpn/checker2-go/internal/module/auth/domain"
	"github.com/hidessh99/sub-cf-vpn/checker2-go/internal/shared/port"
	"github.com/hidessh99/sub-cf-vpn/checker2-go/pkg/apperror"
	"github.com/hidessh99/sub-cf-vpn/checker2-go/pkg/jwt"
	"github.com/hidessh99/sub-cf-vpn/checker2-go/pkg/password"
)

type AuthUseCase interface {
	Login(ctx context.Context, username, passwordPlain string) (string, *domain.Admin, error)
	GetProfile(ctx context.Context, adminID uint) (*domain.Admin, error)
	ChangePassword(ctx context.Context, adminID uint, oldPasswordPlain, newPasswordPlain string) error
}

type authUseCase struct {
	adminRepo domain.AdminRepository
	cfg       *config.AppConfig
	log       port.Logger
}

func NewAuthUseCase(adminRepo domain.AdminRepository, cfg *config.AppConfig, log port.Logger) AuthUseCase {
	return &authUseCase{
		adminRepo: adminRepo,
		cfg:       cfg,
		log:       log,
	}
}

func (u *authUseCase) Login(ctx context.Context, username, passwordPlain string) (string, *domain.Admin, error) {
	admin, err := u.adminRepo.FindByUsername(ctx, username)
	if err != nil {
		u.log.Error("Database error during username search", err, "AuthUseCase")
		return "", nil, err
	}

	if admin == nil {
		u.log.Warn(fmt.Sprintf("Login failed - user not found: %s", username), "AuthUseCase")
		return "", nil, apperror.NewUnauthorizedError("Invalid username or password")
	}

	err = password.VerifyPassword(passwordPlain, admin.Password)
	if err != nil {
		u.log.Warn(fmt.Sprintf("Login failed - invalid password for user: %s", username), "AuthUseCase")
		return "", nil, apperror.NewUnauthorizedError("Invalid username or password")
	}

	// Generate JWT token
	claims := map[string]interface{}{
		"id":       admin.ID,
		"username": admin.Username,
	}

	token, err := jwt.SignToken(claims, u.cfg.JWT.Secret, u.cfg.JWT.ExpiresIn)
	if err != nil {
		u.log.Error("Failed to sign JWT token", err, "AuthUseCase")
		return "", nil, err
	}

	// Clear password in returning model
	admin.Password = ""

	return token, admin, nil
}

func (u *authUseCase) GetProfile(ctx context.Context, adminID uint) (*domain.Admin, error) {
	u.log.Debug(fmt.Sprintf("Fetching profile for admin id=%d", adminID), "AuthUseCase")

	admin, err := u.adminRepo.FindByID(ctx, adminID)
	if err != nil {
		u.log.Error(fmt.Sprintf("Database error during profile retrieval for admin id=%d", adminID), err, "AuthUseCase")
		return nil, err
	}

	if admin == nil {
		u.log.Warn(fmt.Sprintf("Profile not found for admin id=%d", adminID), "AuthUseCase")
		return nil, apperror.NewNotFoundError("Admin not found")
	}

	// Clear password in returning model
	admin.Password = ""

	return admin, nil
}

func (u *authUseCase) ChangePassword(ctx context.Context, adminID uint, oldPasswordPlain, newPasswordPlain string) error {
	admin, err := u.adminRepo.FindByID(ctx, adminID)
	if err != nil {
		u.log.Error(fmt.Sprintf("Database error during profile fetch for password change, admin id=%d", adminID), err, "AuthUseCase")
		return err
	}

	if admin == nil {
		u.log.Warn(fmt.Sprintf("Password change failed - admin not found: id=%d", adminID), "AuthUseCase")
		return apperror.NewNotFoundError("Admin not found")
	}

	err = password.VerifyPassword(oldPasswordPlain, admin.Password)
	if err != nil {
		u.log.Warn(fmt.Sprintf("Password change failed - incorrect current password for admin id=%d", adminID), "AuthUseCase")
		return apperror.NewValidationError("Incorrect current password")
	}

	if len(newPasswordPlain) < 6 {
		return apperror.NewValidationError("New password must be at least 6 characters long")
	}

	hashed, err := password.HashPassword(newPasswordPlain)
	if err != nil {
		u.log.Error("Failed to hash new password during password change", err, "AuthUseCase")
		return err
	}

	err = u.adminRepo.UpdatePassword(ctx, adminID, hashed)
	if err != nil {
		u.log.Error(fmt.Sprintf("Database error during password update for admin id=%d", adminID), err, "AuthUseCase")
		return err
	}

	return nil
}
